server/plugin.go

```diff
@@ -113,36 +118,76 @@ func (p *Plugin) handleEvent(ev model.PluginClusterEvent) error {
 		return err
 	}
 
-	p.mut.RLock()
-	us := p.sessions[msg.ConnID]
-	p.mut.RUnlock()
-
 	switch clusterMessageType(ev.Id) {
 	case clusterMessageTypeConnect:
+		p.LogDebug("connect event", "ChannelID", msg.ChannelID, "UserID", msg.UserID, "ConnID", msg.ConnID)
+		p.mut.Lock()
+		defer p.mut.Unlock()
+		us := p.sessions[msg.ConnID]
 		if us != nil {
 			return fmt.Errorf("session already exists, userID=%q, connID=%q, channelID=%q",
 				us.userID, msg.ConnID, us.channelID)
 		}
-		us := newUserSession(msg.UserID, msg.ChannelID, msg.ConnID)
-		p.mut.Lock()
+		us = newUserSession(msg.UserID, msg.ChannelID, msg.ConnID, true)
 		p.sessions[msg.ConnID] = us
-		p.mut.Unlock()
 		go p.startSession(us, msg.SenderID)
+		return nil
+	case clusterMessageTypeReconnect:
+		p.LogDebug("reconnect event", "UserID", msg.UserID, "ConnID", msg.ConnID)
+
+		p.mut.Lock()
+		defer p.mut.Unlock()
+
+		us := p.sessions[msg.ConnID]
+		if us == nil {
+			return nil
+		}
+
+		if atomic.CompareAndSwapInt32(&us.wsReconnected, 0, 1) {
+			p.LogDebug("closing reconnectCh", "connID", msg.ConnID)
+			close(us.wsReconnectCh)
```

Comment:
cool -- had to follow this through, makes sense now.

Reply:
Unfortunately overall logic is not very simple but there's a lot to consider there and had to cover for potential misuses as well (e.g. clients sending multiple/out of order events). Happy to answer anything that's not clear.